### ISSUE TITLE
Stabilization metabug: 1.0-alpha

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#19260]
This is a tracking metabug for stabilization of both APIs and gated features.

### First pass stabilization/reform

For simpler/mostly stable APIs (core types, iterators), stabilization is mostly a matter of aligning with conventions and deprecating marginal methods and takes place directly under the supervision of the core team. For deeper redesign, RFCs are needed. Where applicable, the list below points to the relevant RFC. 

Most APIs with a chechmark are still `#unstable`, either due to developing conventions, lack of language features (e.g. unboxed closures), or some other problem. We plan to do a rapid second pass closer to the 1.0.0 beta.

(You can track the following list [quantitatively](http://static.rust-lang.org/doc/master/std/stability.html), though the fact that trait impls count exaggerates the amount of `#[experimental]` items.)

**Items are marked complete when a PR/RFC has been posted** -- this is just tracking the design work.

* [x] `any` - #15936 
* [x] `ascii` (see [RFC](https://github.com/rust-lang/rfcs/pull/486))
* [x] `bitflags` (see [RFC](https://github.com/rust-lang/rfcs/pull/509))
* [x] `bool` - #15936
* [x] `borrow` (see [RFC](https://github.com/rust-lang/rfcs/pull/235))
* [x] `boxed` - #15591
* [x] `c_str` (see [RFC](https://github.com/rust-lang/rfcs/pull/494))
* [x] `c_vec` (see [RFC](https://github.com/rust-lang/rfcs/pull/494))
* [x] `cell` - #15591
* [x] `char` - #18603
* [x] `clone` - #15135
* [x] `cmp` - #15797 (see [RFC](https://github.com/rust-lang/rfcs/pull/439))
* [x] `collections` (see [RFC](https://github.com/rust-lang/rfcs/pull/235))
* [x] `comm` - #15286
* [x] `default` - #15797
* [x] `dynamic_lib` (`#[unstable]`)
* [x] `error` (see [RFC](https://github.com/rust-lang/rfcs/pull/201))
* [x] `f32`, `f64` (see [RFC](https://github.com/rust-lang/rfcs/pull/369))
* [x] `finally` (`#[unstable]`)
* [x] `fmt` (see [RFC](https://github.com/rust-lang/rfcs/pull/380))
* [x] `hash` - #19673
* [x] `i8`, `i16`, `i32`, `i64`, `int` (see [RFC](https://github.com/rust-lang/rfcs/pull/369))
* [x] `intrinsics` (`#[unstable]`)
* [x] `io` (see [RFC](https://github.com/rust-lang/rfcs/pull/517))
* [x] `iter` - #19176
* [x] `kinds` (punted to second pass)
* [x] `thread_local` (see [RFC](https://github.com/rust-lang/rfcs/pull/461))
* [x] `macros` (punted to second pass)
* [x] `mem` - #14392
* [x] `num` (see [RFC](https://github.com/rust-lang/rfcs/pull/369))
* [x] `ops` (see [RFC](https://github.com/rust-lang/rfcs/pull/439))
* [x] `option` - #16664
* [x] `os` (see [RFC](https://github.com/rust-lang/rfcs/pull/517))
* [x] `path` (see [RFC](https://github.com/rust-lang/rfcs/pull/474))
* [x] `prelude` (see [RFC](https://github.com/rust-lang/rfcs/pull/503))
* [x] `rand` (`#[unstable]`)
* [x] `raw` (`#[unstable]`)
* [x] `rc` - #15797
* [x] `result` - #16664
* [x] `rt` (`#[unstable]`)
* [x] `simd` (`#[unstable]`)
* [x] `slice` - #16332
* [x] `str` - #19741, (see [RFC](https://github.com/rust-lang/rfcs/pull/528))
* [x] `string` - #17438
* [x] `sync` - #19274
* [x] `task` (see [RFC](https://github.com/rust-lang/rfcs/pull/230))
* [x] `time` (punted to third pass)
* [x] `tuple` - (likely to be deprecated before 1.0), #15936
* [x] `u8`, `u16`, `u32`, `u64`, `uint` (see [RFC](https://github.com/rust-lang/rfcs/pull/369))
* [x] `unit` - #15936
* [x] `vec` - #17029

### Second pass stabilization: for alpha

The second pass is about actually moving APIs to `#[stable]` status; it may involve a bit of minor shuffling, but any major design work should have already taken place. In many cases APIs were blocked on language features (like unboxed closures) before moving to this status.

* [x] ~~`any`~~ (`#[unstable` for alpha)
* [x] ~~`ascii`~~ (`#[unstable]` for alpha)
* [x] ~~`bitflags`~~ (see [RFC](https://github.com/rust-lang/rfcs/blob/master/text/0509-collections-reform-part-2.md))
* [x] `bool` - acrichto #20006
* [x] ~~`borrow`~~ (`#[unstable]` for alpha)
* [x] `boxed` - acrichto #20420
* [x] `c_str` - acrichto #20507
* [x] `c_vec` - acrichto #20507
* [x] `cell` - aturon #20066 -- `std` module mistakenly left experimental
* [x] `char` - huonw #20395
* [x] `clone` - aturon #20070
* [x] `cmp` - aturon #20065
* [x] `collections` - aturon / @Gankro #20356
* [x] `comm` - acrichto #20273
* [x] `default` - acrichto #19902
* [x] ~~`dynamic_lib`~~ (`#[unstable]`)
* [x] `error` - aturon #20647
* [x] `f32`, `f64` - aturon (note: already stable; the `num::Float`/`FloatMath` traits are the problem point)
* [x] ~~`finally`~~ (`#[unstable]`)
* [ ] `fmt` - acrichto
* [x] `hash` - acrichto #20654
* [x] `i8`, `i16`, `i32`, `i64`, `int` - aturon #20062
* [x] ~~`intrinsics`~~ (`#[unstable]` for alpha)
* [x] ~~`io`~~ (`#[unstable]` for alpha)
* [x] `iter` - aturon #20560
* [x] `kinds` - aturon/nrc, flatten to `markers` #20607
* [x] `macros` - acrichto #20657
* [x] `mem` - acrichto #19902
* [x] `num` - aturon/huonw #20573
* [x] `ops` - aturon #20560
* [x] `option` - acrichto #19849
* [x] ~~`os`~~ (`#[unstable]` for alpha)
* [x] ~~`path`~~ (`#[unstable]` for alpha)
* [x] `prelude` - acrichto #20157
* [x] `ptr` - acrichto #20042
* [x] ~~`rand`~~ (`#[unstable]` for alpha)
* [x] ~~`raw`~~ (`#[unstable]` for alpha)
* [x] `rc` - aturon (complete from pass 1)
* [x] `result` - acrichto #19961
* [x] ~~`rt`~~ (`#[unstable]`)
* [x] ~~`simd`~~ (`#[unstable]`)
* [x] `slice` - aturon #20061, #20560
* [x] `str` - acrichto #19741 / kimundi #20058
* [x] `string` - acrichto #20306
* [x] `sync` - acrichto #19661, #20315
* [x] `task` (`#[deprecated]`)
* [x] `thread` - aturon #20615
* [x] `thread_local` - acrichto #20354
* [x] ~~`thunk`~~ (`#[unstable]`)
* [x] `tuple` - acrichto #20006
* [x] `u8`, `u16`, `u32`, `u64`, `uint` - aturon #20062
* [x] `unit` - acrichto #20006
* [x] `vec` - aturon #20061

### RFCs for reform

The following RFCs tackle deeper API redesigns:

#### Accepted

* [Error interoperation](https://github.com/rust-lang/rfcs/pull/201)
* [Runtime reform](https://github.com/rust-lang/rfcs/pull/230)
* [Collections reform](https://github.com/rust-lang/rfcs/pull/235) 
* [Numerics reform](https://github.com/rust-lang/rfcs/pull/369)
* [Formatting stabilization](https://github.com/rust-lang/rfcs/pull/380) 
* [cmp/ops reform](https://github.com/rust-lang/rfcs/pull/439)
* [Path reform](https://github.com/rust-lang/rfcs/pull/474)
* [Thread-local storage reform](https://github.com/rust-lang/rfcs/pull/461)
* [Split `Show`](https://github.com/rust-lang/rfcs/pull/504)
* [Prelude stabilization](https://github.com/rust-lang/rfcs/pull/503)
* [Collections reform part 2](https://github.com/rust-lang/rfcs/pull/509)
* [Macro reform](https://github.com/rust-lang/rfcs/pull/453)
* [c_str, c_vec reform](https://github.com/rust-lang/rfcs/pull/494)

#### Pending

* [io and os reform](https://github.com/rust-lang/rfcs/pull/517/)

## Non-`std` crates

See https://github.com/rust-lang/rust/issues/18585#issuecomment-61560798

### Just need `#[experimental]`

We will ship these crates, but will they will participate in `#![staged_api]`
and will not be accessible in the stable channel.

* [x] alloc
* [x] arena
* [x] collections
* [x] core
* [x] flate
* [x] fmt_macros
* [x] graphviz
* [x] rand
* [x] rbml
* [x] rustc
* [x] rustc_back
* [x] rustc_llvm
* [x] rustc_trans
* [x] rustdoc
* [x] syntax
* [x] unicode

### Need `Cargo.toml` and a publishing strategy

These crates will all be distributed, but participate in `#![staged_api]`. This
means they will not be accessible in the stable channel's standard
distribution, but they will all be available through crates.io.

* [x] getopts - https://github.com/rust-lang/getopts
* [x] libc - #20606
* [x] log - https://github.com/rust-lang/log
* [x] regex - https://github.com/rust-lang/regex
* [x] regex_macros - https://github.com/rust-lang/regex
* [x] term - https://github.com/rust-lang/term

### Need cooperation with the standard library due to compiler internals

* [x] serialize - #19755, https://github.com/rust-lang/rustc-serialize
* [x] test - #20603

### Removed or to remove

* [x] green - #18967
* [x] native - #18967
* [x] rustrt - #19654
* [x] sync - #19255
* [x] time - #18858

### Ungating

* [x] Globs
* [x] Tuple indexing ([RFC](https://github.com/rust-lang/rfcs/pull/450)), #19472
* [x] Struct variants ([RFC](https://github.com/rust-lang/rfcs/pull/418)), #18994
* [x] `if let`/`while let` ([RFC](https://github.com/rust-lang/rfcs/pull/450)), #19472
* [x] Macros
* [x] Associated types
* [x] Default type parameters

### Removal

* [x] Slicing syntax: should be removed in favor of [`ops` reform](https://github.com/rust-lang/rfcs/pull/439)

## CLI Tooling

* [ ] command line flags - #19051, #19900

**Comment by aturon** [comment]
I've renamed this to stabilization for 1.0-alpha, which is complete. There's a [new issue](https://github.com/rust-lang/rust/issues/20761) for 1.0-beta.
